File: src/main/java/com/refinedmods/refinedstorage/network/grid/PortableGridItemUpdateMessage.java
package com.refinedmods.refinedstorage.network.grid;

import com.refinedmods.refinedstorage.api.util.StackListEntry;
import com.refinedmods.refinedstorage.screen.BaseScreen;
import com.refinedmods.refinedstorage.screen.grid.GridScreen;
import com.refinedmods.refinedstorage.screen.grid.stack.IGridStack;
import com.refinedmods.refinedstorage.screen.grid.view.GridViewImpl;
import com.refinedmods.refinedstorage.blockentity.grid.portable.IPortableGrid;
import com.refinedmods.refinedstorage.util.StackUtils;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.world.item.ItemStack;
import net.minecraftforge.network.NetworkEvent;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Supplier;

public class PortableGridItemUpdateMessage {
    private IPortableGrid portableGrid;

    private List<IGridStack> stacks = new ArrayList<>();

    public PortableGridItemUpdateMessage(List<IGridStack> stacks) {
        this.stacks = stacks;
    }

    public PortableGridItemUpdateMessage(IPortableGrid portableGrid) {
        this.portableGrid = portableGrid;
    }

    public static PortableGridItemUpdateMessage decode(FriendlyByteBuf buf) {
        int size = buf.readInt();

        List<IGridStack> stacks = new ArrayList<>();

        for (int i = 0; i < size; ++i) {
            stacks.add(StackUtils.readItemGridStack(buf));
        }

        return new PortableGridItemUpdateMessage(stacks);
    }

    public static void encode(PortableGridItemUpdateMessage message, FriendlyByteBuf buf) {
        int size = message.portableGrid.getItemCache().getList().getStacks().size();

        buf.writeInt(size);

        for (StackListEntry<ItemStack> stack : message.portableGrid.getItemCache().getList().getStacks()) {
            StackUtils.writeItemGridStack(buf, stack.getStack(), stack.getId(), null, false, message.portableGrid.getItemStorageTracker().get(stack.getStack()));
        }
    }

    public static void handle(PortableGridItemUpdateMessage message, Supplier<NetworkEvent.Context> ctx) {
        BaseScreen.executeLater(GridScreen.class, grid -> {
            grid.setView(new GridViewImpl(grid, GridScreen.getDefaultSorter(), GridScreen.getSorters()));
            grid.getView().setStacks(message.stacks);
            grid.getView().sort();
        });

        ctx.get().setPacketHandled(true);
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/network/grid/PortableGridItemUpdateMessage.kt
package com.refinedmods.refinedstorage.network.grid

import com.refinedmods.refinedstorage.screen.BaseScreen
import com.refinedmods.refinedstorage.screen.grid.GridScreen
import com.refinedmods.refinedstorage.screen.grid.stack.IGridStack
import com.refinedmods.refinedstorage.screen.grid.view.ItemGridView
import com.refinedmods.refinedstorage.tile.grid.portable.IPortableGrid
import com.refinedmods.refinedstorage.util.StackUtils.readItemGridStack
import com.refinedmods.refinedstorage.util.StackUtils.writeItemGridStack
import net.minecraft.network.PacketByteBuf
import net.minecraftforge.fml.network.NetworkEvent
import java.util.*
import java.util.function.Supplier

class PortableGridItemUpdateMessage {
    private var portableGrid: IPortableGrid? = null
    private var stacks: List<IGridStack> = ArrayList()

    constructor(stacks: List<IGridStack>) {
        this.stacks = stacks
    }

    constructor(portableGrid: IPortableGrid?) {
        this.portableGrid = portableGrid
    }

    companion object {
        fun decode(buf: PacketByteBuf): PortableGridItemUpdateMessage {
            val size: Int = buf.readInt()
            val stacks: MutableList<IGridStack> = ArrayList()
            for (i in 0 until size) {
                stacks.add(readItemGridStack(buf))
            }
            return PortableGridItemUpdateMessage(stacks)
        }

        fun encode(message: PortableGridItemUpdateMessage, buf: PacketByteBuf) {
            val size: Int = message.portableGrid!!.itemCache.getList().getStacks().size()
            buf.writeInt(size)
            for (stack in message.portableGrid!!.itemCache.getList().getStacks()) {
                writeItemGridStack(buf, stack.stack!!, stack.id, null, false, message.portableGrid!!.itemStorageTracker[stack.stack])
            }
        }

        fun handle(message: PortableGridItemUpdateMessage, ctx: Supplier<NetworkEvent.Context>) {
            BaseScreen.executeLater(GridScreen::class.java) { grid: GridScreen ->
                grid.view = ItemGridView(grid, GridScreen.defaultSorter, GridScreen.sorters)
                grid.view.stacks = message.stacks
                grid.view.sort()
            }
            ctx.get().setPacketHandled(true)
        }
    }
}